cmd/ztest/ztest.c

```diff
@@ -7052,6 +7119,379 @@ ztest_import(ztest_shared_t *zs)
+	 * restarting test.
+	 */
+	ztest_shared_opts->zo_raidz_expand_test = UINT64_MAX; /* pass 1 done */
+	if (ztest_opts.zo_verbose >= 1) {
+		(void) printf("raidz expandsion reflow started,"
+		    " waiting for offset %llu to be reched\n",
+		    (u_longlong_t)desreflow);
+	}
+
+	while (pres->pres_reflowed < desreflow) {
+		txg_wait_synced(spa_get_dsl(spa), 0);
+		(void) poll(NULL, 0, 100); /* wait 1/10 second */
+		spa_config_enter(spa, SCL_CONFIG, FTAG, RW_READER);
+		(void) spa_raidz_expand_get_stats(spa, pres);
+		spa_config_exit(spa, SCL_CONFIG, FTAG);
+	}
+
+	/*
+	 * XXX - should we clear the reflow pause here?
+	 */
+	if (ztest_opts.zo_verbose >= 1) {
+		(void) printf("killing raidz expandsion test offset at %llu\n",
+		    (u_longlong_t)pres->pres_reflowed);
+	}
+	ztest_kill(zs);
+
+done:
+#if 0
+	/*
+	 * XXX - we don't have any threads here because we killed the test and
+	 * then ran a scrub in this thread rather than starting sub threads.
+	 */
+	/*
+	 * Wait for all of the tests to complete.
+	 */
+	for (t = 0; t < ztest_opts.zo_threads; t++)
+		VERIFY0(thread_join(run_threads[t]));
+
+	/*
+	 * Close all datasets. This must be done after all the threads
+	 * are joined so we can be sure none of the datasets are in-use
+	 * by any of the threads.
+	 */
+	for (t = 0; t < ztest_opts.zo_threads; t++) {
+		if (t < ztest_opts.zo_datasets)
+			ztest_dataset_close(t);
+	}
+#endif
+
+	txg_wait_synced(spa_get_dsl(spa), 0);
+
+	zs->zs_alloc = metaslab_class_get_alloc(spa_normal_class(spa));
+	zs->zs_space = metaslab_class_get_space(spa_normal_class(spa));
+
+	umem_free(run_threads, ztest_opts.zo_threads * sizeof (kthread_t *));
+
+	/* Kill the resume and deadman threads */
+	ztest_exiting = B_TRUE;
+	VERIFY0(thread_join(resume_thread));
+	VERIFY0(thread_join(deadman_thread));
+	ztest_resume(spa);
+
+	/*
+	 * Right before closing the pool, kick off a bunch of async I/O;
+	 * spa_close() should wait for it to complete.
+	 */
+	for (object = 1; object < 50; object++) {
+		dmu_prefetch(spa->spa_meta_objset, object, 0, 0, 1ULL << 20,
+		    ZIO_PRIORITY_SYNC_READ);
+	}
+
+	/* Verify that at least one commit cb was called in a timely fashion */
+	if (zc_cb_counter >= ZTEST_COMMIT_CB_MIN_REG)
+		VERIFY0(zc_min_txg_delay);
+
+	spa_close(spa, FTAG);
+
+	/*
+	 * Verify that we can loop over all pools.
+	 */
+	mutex_enter(&spa_namespace_lock);
+	for (spa = spa_next(NULL); spa != NULL; spa = spa_next(spa))
+		if (ztest_opts.zo_verbose > 3)
+			(void) printf("spa_next: found %s\n", spa_name(spa));
+	mutex_exit(&spa_namespace_lock);
+
+	/*
+	 * Verify that we can export the pool and reimport it under a
+	 * different name.
+	 */
+	if ((ztest_random(2) == 0) && !ztest_opts.zo_mmp_test) {
+		char name[ZFS_MAX_DATASET_NAME_LEN];
+		(void) snprintf(name, sizeof (name), "%s_import",
+		    ztest_opts.zo_pool);
+		ztest_spa_import_export(ztest_opts.zo_pool, name);
+		ztest_spa_import_export(name, ztest_opts.zo_pool);
+	}
```

Comment:
looks like this was copied from the ztest_run(), but I'm not sure that all of these checks are especially relevant for expansion.  Alternatively, we might break up ztest_run() into 3 sections, allowing us to reuse the code:

1. Set up shared state, open pool, create deadman thread
2. run threads (this would be different for regular vs expansion tests)
3. final checks and close pool

Reply:
Again, here from copying ztest_run and hacking it for our needs will remove if not useful for expansion test.  Left it in as it just seemed like it might provide an additional sanity check.